Pydantic v2 native implementation

febus982 opened the pull request:
Fixes https://github.com/cloudevents/sdk-python/issues/215
Iterates over https://github.com/cloudevents/sdk-python/pull/218

## Changes

Pydantic V2 implementation of CloudEvent class:
- Importing the class/functions from `cloudevents.pydantic` module will import the correct version
- It's possible to import class/functions from `cloudevents.pydantic.v1` or `cloudevents.pydantic.v2` modules. (There's no check on users with V1 deciding to import the V2 module)
- Tests are parametric on both the implementations

## One line description for the changelog

Pydantic V2 implementation of CloudEvent class 

- [x] Tests pass
- [ ] Appropriate changes to README are included in PR

febus982 commented on cloudevents/pydantic/__init__.py:
```diff
@@ -11,7 +11,29 @@
+        from cloudevents.pydantic.v1.event import CloudEvent
+
+    else:
+        from cloudevents.pydantic.v2.conversion import (  # type: ignore
```
Same typing error as previous pull request due to re-assignment of different objects 

https://github.com/cloudevents/sdk-python/pull/218/files#diff-3feb86f607bffb91257e35096edfa40b2d6a291cdc66d404dc139c71531e8b1bR27

febus982 commented on cloudevents/tests/test_pydantic_cloudevent.py:
```diff
@@ -140,29 +159,44 @@ def test_http_cloudevent_mutates_equality(
 
-def test_cloudevent_missing_specversion():
+def test_cloudevent_missing_specversion(cloudevents_implementation):
+    errors = {
+        "v1": "value is not a valid enumeration member; permitted: '0.3', '1.0'",
+        "v2": "Input should be '0.3' or '1.0'",
+    }
```
I'm not sure we should check the error string representation. I left it for consistency.

febus982 commented on requirements/test.txt:
```diff
@@ -10,4 +10,4 @@ aiohttp
 requests
 flask
-pydantic>=1.0.0,<3.0
+pydantic>=2.0.0,<3.0
```
We want to make sure to run the tests using pydantic v2 or the new implementation tests will fail

febus982 commented on cloudevents/tests/test_pydantic_conversions.py:
```diff
@@ -146,12 +176,15 @@ def test_from_dict():
+    assert cloudevents_implementation["from_dict"](given).dict() == given
 
 
+@pytest.mark.xfail  # Pydantic V2 doesn't support function params
```
As the comment says, pydantic V2 doesn't accept kwargs on the json() method anymore.

I haven't found any usage of event.json() in this SDK, I wonder if we should test this at all.

febus982 commented on cloudevents/pydantic/v2/conversion.py:
```diff
@@ -0,0 +1,75 @@
+#  Copyright 2018-Present The CloudEvents Authors
```
All these functions are the same as the V1 ones, I thought about using the same V1/V2 detection logic as the __init__.py file.

If we want to keep the ability for pydantic V2 users to choose the V1 implementation anyway, we need to keep them separate and duplicate.

febus982 commented on cloudevents/pydantic/v2/event.py:
```diff
@@ -0,0 +1,222 @@
+        :return: Event serialized as a standard CloudEvent dict with user specific
+        parameters.
+        """
+        model_dump: dict = handler(self)
+        if isinstance(self.data, (bytes, bytearray, memoryview)):
+            model_dump["data_base64"] = base64.b64encode(self.data).decode("ascii")
+        return model_dump
```
If data is a nested BaseModel, this logic could produce a non compliant JSON string.

Do we assume that when serializing to JSON the contentdatatype is `application/json`? Then `data` should be serialized in a JSON string before serializing the model.

The concept of assumption sounds wrong, though. Perhaps I can add a `serialize_data` method that can be overridden, or, even better a small dictionary that acts as a registry (so people can create other base classes supporting multiple contentdatatypes.

```python
def json_data_serializer(data):
    return json_serialized_data
    
class CloudEvent(abstract.CloudEvent, BaseModel):
    serializer_registry = {
        "application/json": json_data_serializer
    }
    
    ...
```

febus982 replied:
I answered this in the previous comment. I'll address this function as well.

febus982 commented:
@xSAVIKx Here it is! I left comments in the parts with non-trivial changes.

One of them is a question for the last tiny piece of code still missing.

xSAVIKx reviewed (commented):
@febus982 thx for the PR. Please take a look the comments noted below.

febus982 commented on cloudevents/pydantic/v2/event.py:
```diff
@@ -0,0 +1,222 @@
+        },
+    )
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_base64_data_input(cls, data: typing.Any) -> typing.Any:
+        """Populates the `data` property if the model gets created using `data_base64`.
+
+        :param data: Input data.
+
+        :return: Event serialized as a standard CloudEvent dict with user specific
+        parameters.
+        """
+        if isinstance(data, dict) and data.get("data_base64") is not None:
+            data["data"] = base64.b64decode(data["data_base64"])
+            del data["data_base64"]
+        return data
```
**Why these 2 functions are here:**

This is handled if only when serialising the event using `cloudevents.conversion.to_structured` function.

The original pydantic v1 implementation had tests covering both serialisation and json deserialisation without passing through the `conversion` module but using the two pydantic functions:
- `Basemodel.parse_raw()` to deserialise from a json string.
- `Basemodel.json()` to serialize to a json string.

I suspect it was done to give support HTTP frameworks using pydantic serialisation logic (e.g. FastAPI) without explicitly calling the `conversion.*` functions.

This function, and the following one (`serialize_model`) are to support this scenario.

=====

**Reasoning about the implementation:**

The original pydantic v1 serialisation implementation (`_ce_json_loads` and `_ce_json_dumps` functions in the V1 implementation) was relying on the http module conversion logic.

I did rewrite the functions decoupling them from the http module as it seemed an anti-pattern, and using `json.dumps` would have worse performance if compared to the pydantic v2 json serializer.

I see now it doesn't make much sense to have a duplicate implementation and it's better to rely on the existing `cloudevents.conversion` module, so we have a single implementation. I'll address this.

xSAVIKx replied:
So yes, the reasoning for using `_ce_` implementations is to make sure the behavior is consistent. We may provide "extra" APIs that are not part of the SDK for advanced/customized use cases, but IMO the standard conversion should be the default one.

febus982 replied:
I have addressed both serialisation and deserialization. It now uses the methods from `conversion` module.

febus982 commented on cloudevents/pydantic/__init__.py:
```diff
@@ -11,7 +11,30 @@
+    from pydantic import VERSION as PYDANTIC_VERSION
+
+    pydantic_major_version = PYDANTIC_VERSION.split(".")[0]
+    if pydantic_major_version == "2":
```
You're right. I'll also add a test to make sure this is correct.

febus982 replied:
I have not been able to test this but it is fixed.

febus982 commented:
Hey @xSAVIKx I addressed all the comments (I left them unresolved so you can double-check).

I don't think the typing error can be addressed because `json.loads` returns `Any` but we know well we're getting a dictionary there. I could add a check on the result being an instance of dict but I think adding a `# type: ignore` is safe enough. I'm also not sure why `self` is identified with a type and not an object, it might be a mypy issue. Let mw know if you have a better recommendation for this.

xSAVIKx commented on cloudevents/pydantic/v1/__init__.py:
```diff
@@ -0,0 +1,18 @@
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from .conversion import from_dict, from_http, from_json
+from .event import CloudEvent
```
Can you please use fully-qualified imports instead?

xSAVIKx commented on cloudevents/pydantic/v2/__init__.py:
```diff
@@ -0,0 +1,18 @@
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from .conversion import from_dict, from_http, from_json
+from .event import CloudEvent
+
+__all__ = ["CloudEvent", "from_json", "from_dict", "from_http"]
```
Same here, please use fully-qualified imports

xSAVIKx reviewed (commented):
@febus982 overall LGTM. There's some pre-commit thing failing, will you take a look? Also, if'd like I can jump in and fix the thing as well as my comments.

febus982 commented:
@xSAVIKx Ready to be merged, all comments addressed.

benvdh-incentro commented:
@xSAVIKx I noticed there hasn't been a release of the package since january 2023, though some very useful PRs (among them this one) got merged to main. I noticed the CI also fails on the fact that `1.9.0` already exists on PyPI, so likely it's just a version bump that's required.

So my question is: are there any plans to get this code released to PyPI ?

xSAVIKx commented:
> @xSAVIKx I noticed there hasn't been a release of the package since january 2023, though some very useful PRs (among them this one) got merged to main. I noticed the CI also fails on the fact that `1.9.0` already exists on PyPI, so likely it's just a version bump that's required.
> 
> So my question is: are there any plans to get this code released to PyPI ?

Hey @benvdh-incentro, sure thing 🙂 I was holding the release till this particular PR is implemented and we should get one probably on the weekend or early next week
